core/state/state_object.go

```diff
@@ -245,10 +251,34 @@ func (self *StateObject) Value() *big.Int {
 	panic("Value on StateObject should never be called")
 }
 
+func (self *StateObject) ForEachStorage(full bool, cb func(key, value common.Hash) bool) {
+	// When iterating over the storage check the cache first
+	for h, sv := range self.storage {
+		if sv.dirty || full {
+			cb(h, sv.value)
+		}
+	}
+
+	// if full storage retrieval is disabled abort here
+	if !full {
+		return
+	}
+
+	it := self.trie.Iterator()
+	for it.Next() {
+		// ignore cached values
+		key := common.BytesToHash(self.trie.GetKey(it.Key))
+		if _, ok := self.storage[key]; !ok {
+			cb(key, common.BytesToHash(it.Value))
+		}
+	}
+}
```

Comment:
Question: since this method depends on the dirty flag, we may not be able to set it to false above. Maybe we could maintain two flags instead? `dirty` to be used by the state object to avoid too many database writes and/or trie ops, and `modified` to track whether we should "diff" it or not?